midi_app_controller/models/app_state.py

```diff
@@ -19,6 +19,8 @@ class AppState(YamlBaseModel):
         Selected MIDI output port.
     recent_binds_for_controller: dict[Path, Optional[Path]]
         Mapping between each controller and which binds setup was last used.
+    recent_midi_ports_for_controller: dict[Path, dict[str, str]]
+        Mapping between controllers and the last used MIDI ports (in/out)
```

Comment:
Either specify that the dict should only have two key (`in`/`out`) or create a new class with these two fields.